python/ray/dag/dag_node_operation.py

```diff
@@ -297,17 +354,96 @@ def _build_dag_node_operation_graph(
             _add_edge(
                 graph[task_idx][_DAGNodeOperationType.WRITE],
                 graph[downstream_task_idx][_DAGNodeOperationType.READ],
+                "nccl"
+                if graph[task_idx][_DAGNodeOperationType.WRITE].requires_nccl
+                else "shm",
             )
     return graph
 
 
+def _node_repr(node: _DAGOperationGraphNode, idx: int, optimized_index: int):
+    """
+    Representation of a node in the visualization of the execution schedule.
+
+    Args:
+        node: The node to be represented.
+        idx: The index of the node in the execution schedule.
+        optimized_index: The index of the node in the optimized execution schedule.
+    """
+    return str(node) + f" {idx},{optimized_index}"
+
+
+def _visualize_execution_schedule(
+    actor_to_execution_schedule: Dict[
+        "ray.actor.ActorHandle", List[_DAGOperationGraphNode]
+    ],
+    actor_to_optimized_schedule: Dict[
+        "ray.actor.ActorHandle", List[_DAGOperationGraphNode]
+    ],
+    graph: Dict[int, Dict[_DAGNodeOperationType, _DAGOperationGraphNode]],
+):
+    """
+    Visualize the execution schedule for each actor.
+
+    Args:
+        actor_to_execution_schedule: A dictionary that maps an actor handle to
+            the execution schedule which is a list of operation nodes.
+        actor_to_optimized_schedule: A dictionary that maps an actor handle to the
+            optimized execution schedule which is a list of operation nodes.
+        graph: A graph where each node is a _DAGOperationGraphNode. The key is
+            `task_idx`, the index to retrieve its task from `idx_to_task`, and
+            the value is a dictionary that maps the _DAGNodeOperationType (READ,
+            COMPUTE, or WRITE) to the corresponding _DAGOperationGraphNode. It is
+            generated by `_build_dag_node_operation_graph`.
+    """
+    try:
+        import graphviz
```

Comment:
Ideally we should have some kind of ascii based viz too, I think this is often more convenient for users and it's possible to test in CI.

Can leave it for a follow-up though.

Reply:
Created https://github.com/ray-project/ray/issues/48190 as a follow up

Reply:
btw, I think ascii based viz only works well with small DAGs, as soon as the DAG becomes medium size, it will be hard to read. I will post an example of graphviz of medium/large DAG.